### PR TITLE
[PrintAsObjC] Disable -Watimport-in-framework-header warning in generated headers

### PR DESCRIPTION
**Pull request by beccadax**
This is a new warning in clang when you use @import in a framework header. That’s often a mistake, but it isn’t in our generated `-Swift.h` headers. If we’re building with a clang that supports this warning, we’ll now emit a pragma to disable it.

Fixes rdar://41552597.

### CODE REVIEW

File: test/PrintAsObjC/imports.swift [hunk 1]
```diff
@@ -5,10 +5,16 @@
 // RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -I %S/Inputs/custom-modules/ -F %S/Inputs/ -parse-as-library %t/imports.swiftmodule -typecheck -emit-objc-header-path %t/imports.h -import-objc-header %S/../Inputs/empty.h -disable-objc-attr-requires-foundation-module
 // RUN: %FileCheck %s < %t/imports.h
 // RUN: %FileCheck -check-prefix=NEGATIVE %s < %t/imports.h
-// RUN: %check-in-clang %t/imports.h -I %S/Inputs/custom-modules/ -F %S/Inputs/
+// RUN: %check-in-clang %t/imports.h -I %S/Inputs/custom-modules/ -F %S/Inputs/ -Watimport-in-framework-header
```
Comment (jrose-apple) [inline test/PrintAsObjC/imports.swift]: Did you verify that the warning fires without your fix here? I think it really does have to look like a framework header for it to show up. We should make sure we're testing the right thing.

Comment (brentdax) [reply test/PrintAsObjC/imports.swift]: You're right, clang looks up path it's compiling for something with a .framework extension. I'll check if one of the other tests is more suitable.

Comment (jrose-apple) [reply test/PrintAsObjC/imports.swift]: It's fine to add a new RUN line, too; sometimes that's lighter weight than adding a whole new test.

File: test/PrintAsObjC/mixed-framework.swift [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
+// RUN: %empty-directory(%t.framework/Headers)
+// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -F %S/Inputs/ -module-name Mixed -import-underlying-module -parse-as-library %s -typecheck -emit-objc-header-path %t.framework/Headers/mixed.h
+// RUN: %FileCheck -check-prefix=CHECK -check-prefix=FRAMEWORK %s < %t.framework/Headers/mixed.h
+// RUN: %check-in-clang -Watimport-in-framework-header -F %S/Inputs/ %t.framework/Headers/mixed.h
 
-// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -module-name Mixed -import-objc-header %S/Inputs/Mixed.framework/Headers/Mixed.h %s -typecheck -emit-objc-header-path %t/mixed-header.h
-// RUN: %FileCheck -check-prefix=CHECK -check-prefix=HEADER %s < %t/mixed-header.h
-// RUN: %check-in-clang -include %S/Inputs/Mixed.framework/Headers/Mixed.h %t/mixed-header.h
+// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -module-name Mixed -import-objc-header %S/Inputs/Mixed.framework/Headers/Mixed.h %s -typecheck -emit-objc-header-path %t.framework/mixed-header.h
+// RUN: %FileCheck -check-prefix=CHECK -check-prefix=HEADER %s < %t.framework/mixed-header.h
```
Comment (jrose-apple) [inline test/PrintAsObjC/mixed-framework.swift]: Nitpick: Did you want to put this in the Headers directory?

Comment (brentdax) [reply test/PrintAsObjC/mixed-framework.swift]: I don't think so, but I could be wrong. My understanding is that the first header it generates is the one that might actually be packaged in a framework, while the second is one that might be generated as a temporary file when using the framework. But I could be misunderstanding the test; I'm not too familiar with why you would or wouldn't pass `-import-underlying-module -parse-as-library` when doing `-emit-objc-header-path`.

Comment (jrose-apple) [reply test/PrintAsObjC/mixed-framework.swift]: There's only one generated header; the first header listed is effectively a bridging header. (If you wanted to closely more emulate a mixed-source framework, you could provide a module map and use `-import-underlying-module` instead. But then only `public` things get printed.)

Oh, and the generated header is always considered part of a framework build product.

Comment (brentdax) [reply test/PrintAsObjC/mixed-framework.swift]: Okay, so we're not really simulating anything very specifically here. Might as well move the header, then.

**Comment by brentdax** [comment]
@swift-ci please smoke test

**Comment by brentdax** [comment]
@jrose-apple Okay with merging this?

**Comment by jrose-apple** [comment]
Ah, yeah, sorry, those were my only remaining comments.

**Review by jrose-apple — APPROVED** [review]

**Comment by jrose-apple** [comment]
Thanks for taking this!

**Comment by brentdax** [comment]
@swift-ci please smoke test and merge

**Comment by brentdax** [comment]
@swift-ci please smoke test linux platform

**Comment by shahmishal** [comment]
@swift-ci please smoke test linux platform

**Pull request merged**